odk2-src/build-scripts.rst

```diff
@@ -0,0 +1,235 @@
+
+.. _build-scripts-internal-settings:
+
+:file:`root/settings.gradle`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This file determines where to look for the :ref:`build-scripts-external`.
+
+The :code:`gradleConfigVersion` corresponds to a tag in the `Gradle Config repository <https://github.com/opendatakit/gradle-config>`_. If the local gradle files are not found, the versions of those files committed under that tag will be downloaded and used.
+
+
+Before downloading those files, this file checks the local :ref:`build-scripts-directory-structure` for gradle-config. If it is found, that is used. Whichever path is chosen, this linkage is established here and made available to all the rest of the gradle files.
+
+This file also looks for library projects in the local directory structure. If they are found, they are built as dependencies. If not, their prebuilt binaries are downloaded.
+
+.. _build-scripts-internal-build:
+
+:file:`root/build.gradle`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This file establishes URLs to use for resolving dependencies. Links to each of the prebuilt binary repositories are included (demo, master, snapshot).
+
+The dependency versions are also managed here.
+
+.. _build-scripts-internal-inner-build:
+
+:file:`root/app/build.gradle`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The file contains the specific build configuration for this project. The ODK 2 projects do not differ greatly from established norms in this configuration. However, many of the constants and version numbers are stored in :ref:`build-scripts-external-variables` and variables are used here. This allows the tools to be upgraded and maintained in unison, and they can be forced to stay in sync.
+
+This file also establishes the product flavors, signing configs, build types, and other standard options found in many Android projects. The unique aspect comes in the :code:`dependencies` block. The different flavors have different dependencies (they will download different prebuilt binaries for their library projects). The demo and snapshot flavors build against the latest from their repositories, while the master flavor is hard coded to a specific version.
+
+.. _build-scripts-external:
+
+External Build Files
+------------------------
+
+These build files are centralized in the `Gradle Config repository <https://github.com/opendatakit/gradle-config>`_. They included shared configuration, versions, and tasks.
+
+.. _build-scripts-external-variables:
+
+:file:`variables.gradle`
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This file contains all the versions and variables strings shared among the projects. Most notably this includes the release code version, the compile targets, the :program:`Java` version, and the composed project build and flavor names.
```

Comment:
This file stores the variant names not the flavor names. The flavor names are in `publish.gradle`.